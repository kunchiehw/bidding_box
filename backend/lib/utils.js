const jwt = require('jsonwebtoken');

const secret = process.env.SHARE_SECRET;
const db = {
  users: {
    wkc: 'password',
    jarron: 'password',
  },
};


function authenticateUser(username, password) {
  return new Promise((resolve, reject) => {
    if (!username || !password || !(username in db.users) || db.users[username] !== password) {
      return reject();
    }

    const token = jwt.sign({ username }, secret, { expiresIn: '30d' });
    return resolve(token);
  });
}
module.exports.authenticateUser = authenticateUser;


function validateJwt(token) {
  return new Promise((resolve, reject) => {
    jwt.verify(token, secret, (err, decoded) => {
      if (err || !decoded) {
        reject();
      }

      if (!('username' in decoded) || !(decoded.username in db.users)) {
        reject();
      }

      resolve(token);
    });
  });
}
module.exports.validateJwt = validateJwt;


module.exports.validateJwtMiddleware = (req, res, next) => {
  if (req.headers.authorization && req.headers.authorization.split(' ')[0] === 'Bearer') {
    const token = req.headers.authorization.split(' ')[1];
    validateJwt(token)
      .then(() => next())
      .catch(() => next(new Error('Unauthorized')));
  } else {
    next(new Error('Unauthorized'));
  }
};


module.exports.getTtl = () =>
  Math.floor(Date.now() / 1000) + (4 * 60 * 60); // ttl for 4 hour
