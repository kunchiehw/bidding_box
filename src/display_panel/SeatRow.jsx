import React from 'react';
import PropTypes from 'prop-types';
import {SEATS, VULS} from '../util/Util.jsx';

const propTypes = {
  vulnerability: PropTypes.oneOf(VULS).isRequired
};

function SeatBlock(seat, vulnerability) {
  const vulnerable = (vulnerability === 'BOTH')
    || (vulnerability === 'NS' && (seat === 'NORTH' || seat === 'SOUTH'))
    || (vulnerability === 'EW' && (seat === 'EAST' || seat === 'WEST'))

  if(vulnerable) {
    return (
      <div key={seat} className='display-block seat-block vulnerable-seat-block'>{seat}</div>
    );
  }
  else {
    return (
      <div key={seat} className='display-block seat-block non-vulnerable-seat-block'>{seat}</div>
    );
  }
}

function SeatRow(props) {
  const SeatBlocks = [];
  for (var seatIndex = 0; seatIndex < SEATS.length; seatIndex++) {
    SeatBlocks.push(SeatBlock(SEATS[seatIndex], props.vulnerability));
  }
  return (
    <div className='seat-row'>
      {SeatBlocks}
    </div>
  );
}

SeatRow.propTypes = propTypes;

export default SeatRow;
