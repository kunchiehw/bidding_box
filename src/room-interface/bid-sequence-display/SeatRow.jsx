import React from 'react';
import PropTypes from 'prop-types';
import { SEATS, SEATS_NS, SEATS_EW, VULS } from '../../util/util';

const propTypes = {
  vulnerability: PropTypes.oneOf(VULS).isRequired,
};

function SeatCell(seat, vulnerability) {
  const vulnerable = (vulnerability === 'BOTH')
    || (vulnerability === 'NS' && SEATS_NS.includes(seat))
    || (vulnerability === 'EW' && SEATS_EW.includes(seat));

  if (vulnerable) {
    return (
      <div key={seat} className="bid-sequence-display-cell seat-cell vulnerable-seat-cell">{seat}</div>
    );
  }

  return (
    <div key={seat} className="bid-sequence-display-cell seat-cell non-vulnerable-seat-cell">{seat}</div>
  );
}

function SeatRow(props) {
  const SeatCells = [];
  for (let seatIndex = 0; seatIndex < SEATS.length; seatIndex += 1) {
    SeatCells.push(SeatCell(SEATS[seatIndex], props.vulnerability));
  }
  return (
    <div className="seat-row">
      {SeatCells}
    </div>
  );
}

SeatRow.propTypes = propTypes;

export default SeatRow;
