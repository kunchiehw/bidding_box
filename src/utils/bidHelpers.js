const { STANDARD_SUITS } = require('./constants');

function isPass(bid) {
  return (bid && bid.suit === 'PASS');
}

function isDouble(bid) {
  return (bid && bid.suit === 'DOUBLE');
}

function isSuit(bid) {
  return (bid && STANDARD_SUITS.includes(bid.suit));
}

function isBidSeqEnded(bidSeq) {
  const bidSeqLen = bidSeq.length;
  if (isPass(bidSeq[bidSeqLen - 3]) &&
      isPass(bidSeq[bidSeqLen - 2]) &&
      isPass(bidSeq[bidSeqLen - 1])) {
    return (bidSeqLen >= 4);
  }
  return false;
}

function isDoubleAllowed(bidSeq) {
  const bidSeqLen = bidSeq.length;
  if (isSuit(bidSeq[bidSeqLen - 1])) {
    return false;
  }
  if (isSuit(bidSeq[bidSeqLen - 3]) &&
      isPass(bidSeq[bidSeqLen - 2]) &&
      isPass(bidSeq[bidSeqLen - 1])) {
    return false;
  }
  return true;
}

function isRedoubleAllowed(bidSeq) {
  const bidSeqLen = bidSeq.length;
  if (isDouble(bidSeq[bidSeqLen - 1])) {
    return false;
  }
  if (isDouble(bidSeq[bidSeqLen - 3]) &&
      isPass(bidSeq[bidSeqLen - 2]) &&
      isPass(bidSeq[bidSeqLen - 1])) {
    return false;
  }
  return true;
}

function getCurrentBid(bidSeq) {
  for (let i = bidSeq.length - 1; i >= 0; i -= 1) {
    if (isSuit(bidSeq[i])) {
      return bidSeq[i];
    }
  }
  return { level: 0, suit: 'PASS' };
}

function isStandardBidAllowed(currentLevel, currentSuit, level, suit) {
  if (currentLevel !== level) {
    return (currentLevel > level);
  }
  return (STANDARD_SUITS.indexOf(currentSuit) >= STANDARD_SUITS.indexOf(suit));
}


module.exports = {
  isPass,
  isDouble,
  isSuit,
  isBidSeqEnded,
  isDoubleAllowed,
  isRedoubleAllowed,
  getCurrentBid,
  isStandardBidAllowed,
};
