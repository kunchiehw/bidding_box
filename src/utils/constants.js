export const POSSIBLE_LEVELS = [1, 2, 3, 4, 5, 6, 7];

export const STANDARD_SUITS = ['CLUBS', 'DIAMONDS', 'HEARTS', 'SPADES', 'NOTRUMPS'];
export const SPECIAL_SUITS = ['PASS', 'DOUBLE', 'REDOUBLE'];
export const ALL_SUITS = ['CLUBS', 'DIAMONDS', 'HEARTS', 'SPADES', 'NOTRUMPS', 'PASS', 'DOUBLE', 'REDOUBLE'];

export const NS_SEATS = ['NORTH', 'SOUTH'];
export const EW_SEATS = ['WEST', 'EAST'];
export const SEATS = ['WEST', 'NORTH', 'EAST', 'SOUTH'];
export const DECLARERS = ['WEST', 'NORTH', 'EAST', 'SOUTH', 'NS', 'EW'];
export const PARTICIPANTS = ['WEST', 'EAST', 'WATCHER'];

export const VULNERABILITIES = ['NONE', 'NS', 'EW', 'BOTH'];

export const BOOLEANS = [true, false];

export const SERVER_API_HOST = `${process.env.REACT_APP_BACKEND_SCHEMA}://${process.env.REACT_APP_BACKEND_URL}`;
export const SERVER_WS_HOST = `ws://${process.env.REACT_APP_BACKEND_URL}`;
