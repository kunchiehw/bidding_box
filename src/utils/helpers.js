import { STANDARD_SUITS } from './constants';
import getMessage from '../i18n/messages';

export function bidToString(level, suit) {
  if (STANDARD_SUITS.includes(suit)) {
    return `${level} ${getMessage(suit)}`;
  }
  return getMessage(suit);
}

export function suitColor(suit) {
  switch (suit) {
    case 'NOTRUMPS':
      return 'black';
    case 'SPADES':
    case 'REDOUBLE':
      return 'blue';
    case 'HEARTS':
    case 'DOUBLE':
      return 'red';
    case 'DIAMONDS':
      return 'orange';
    case 'CLUBS':
    case 'PASS':
      return 'green';
    default:
      return '';
  }
}

export function shouldStandardBidButtonDisabled(currentLevel, currentSuit, buttonLevel, buttonSuit) {
  if (currentLevel !== buttonLevel) {
    return (currentLevel > buttonLevel);
  }
  return (STANDARD_SUITS.indexOf(currentSuit) >= STANDARD_SUITS.indexOf(buttonSuit));
}
