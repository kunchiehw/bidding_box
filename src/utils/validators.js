import { POSSIBLE_LEVELS, STANDARD_SUITS } from './constants';

// HOC for all validator
function createChainableTypeChecker(validate) {
  function checkType(isRequired, props, propName, componentName, location) {
    componentName = componentName || 'ANONYMOUS';
    if (props[propName] == null) {
      if (isRequired) {
        return new Error(`'${componentName}': requires prop '${propName}'.`);
      }
      return null;
    }
    return validate(props, propName, componentName, location);
  }

  const chainedCheckType = checkType.bind(null, false);
  chainedCheckType.isRequired = checkType.bind(null, true);

  return chainedCheckType;
}

export const levelValidator = createChainableTypeChecker((props, propName, componentName) => {
  componentName = componentName || 'ANONYMOUS';

  if (props.suit == null) {
    return new Error(`'${componentName}': requires prop 'suit' when validates prop '${propName}'.`);
  }

  if (props[propName]) {
    const level = props[propName];
    if (STANDARD_SUITS.includes(props.suit)) {
      if (!POSSIBLE_LEVELS.includes(level)) {
        return new Error(`'${componentName}': '${propName}' should be in POSSIBLE_LEVELS when 'suit' is in STANDARD_SUITS.`);
      }
    } else if (level !== 0) {
      return new Error(`'${componentName}': '${propName}' should be 0 when 'suit' is in SPECIAL_SUITS.`);
    }
  }

  return null;
});

export function foo() {}
