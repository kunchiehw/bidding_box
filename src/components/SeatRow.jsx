import React, { PureComponent } from 'react';
import PropTypes from 'prop-types';
import { Grid, Segment } from 'semantic-ui-react';
import { SEATS, VULNERABILITIES } from '../utils/constants';
import { isVulnerable } from '../utils/helpers';

const propTypes = {
  vulnerability: PropTypes.oneOf(VULNERABILITIES).isRequired,
};

class SeatRow extends PureComponent {
  render() {
    return (
      <Grid.Row key="seat" className="seat-row" column={4}>
        {SEATS.map((seat) => {
          const seatIsVulnerable = isVulnerable(seat, this.props.vulnerability);
          return ((
            <Grid.Column key={seat}>
              <Segment
                className="seat-cell"
                color={(seatIsVulnerable) ? 'red' : null}
                inverted={seatIsVulnerable}
              >
                {seat}
              </Segment>
            </Grid.Column>
          ));
        })}
      </Grid.Row>
    );
  }
}

SeatRow.propTypes = propTypes;

export default SeatRow;
