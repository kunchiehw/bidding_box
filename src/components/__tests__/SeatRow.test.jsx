import React from 'react';
import { shallow } from 'enzyme';
import toJson from 'enzyme-to-json';
import SeatRow from '../SeatRow';
import { VULNERABILITIES } from '../../utils/constants';

describe('SeatRow render', () => {
  VULNERABILITIES.forEach((vul) => {
    it(`vulnerability: ${vul}`, () => {
      const seatRow = shallow(<SeatRow
        vulnerability={vul}
      />);
      expect(toJson(seatRow)).toMatchSnapshot();
    });
  });
});
