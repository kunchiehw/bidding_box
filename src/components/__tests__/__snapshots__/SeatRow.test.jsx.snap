// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`SeatRow render vulnerability: BOTH 1`] = `
<GridRow
  className="seat-row"
  column={4}
  key="seat"
>
  <GridColumn
    key="WEST"
  >
    <Segment
      className="seat-cell"
      color="red"
      inverted={true}
    >
      WEST
    </Segment>
  </GridColumn>
  <GridColumn
    key="NORTH"
  >
    <Segment
      className="seat-cell"
      color="red"
      inverted={true}
    >
      NORTH
    </Segment>
  </GridColumn>
  <GridColumn
    key="EAST"
  >
    <Segment
      className="seat-cell"
      color="red"
      inverted={true}
    >
      EAST
    </Segment>
  </GridColumn>
  <GridColumn
    key="SOUTH"
  >
    <Segment
      className="seat-cell"
      color="red"
      inverted={true}
    >
      SOUTH
    </Segment>
  </GridColumn>
</GridRow>
`;

exports[`SeatRow render vulnerability: EW 1`] = `
<GridRow
  className="seat-row"
  column={4}
  key="seat"
>
  <GridColumn
    key="WEST"
  >
    <Segment
      className="seat-cell"
      color="red"
      inverted={true}
    >
      WEST
    </Segment>
  </GridColumn>
  <GridColumn
    key="NORTH"
  >
    <Segment
      className="seat-cell"
      color={null}
      inverted={false}
    >
      NORTH
    </Segment>
  </GridColumn>
  <GridColumn
    key="EAST"
  >
    <Segment
      className="seat-cell"
      color="red"
      inverted={true}
    >
      EAST
    </Segment>
  </GridColumn>
  <GridColumn
    key="SOUTH"
  >
    <Segment
      className="seat-cell"
      color={null}
      inverted={false}
    >
      SOUTH
    </Segment>
  </GridColumn>
</GridRow>
`;

exports[`SeatRow render vulnerability: NONE 1`] = `
<GridRow
  className="seat-row"
  column={4}
  key="seat"
>
  <GridColumn
    key="WEST"
  >
    <Segment
      className="seat-cell"
      color={null}
      inverted={false}
    >
      WEST
    </Segment>
  </GridColumn>
  <GridColumn
    key="NORTH"
  >
    <Segment
      className="seat-cell"
      color={null}
      inverted={false}
    >
      NORTH
    </Segment>
  </GridColumn>
  <GridColumn
    key="EAST"
  >
    <Segment
      className="seat-cell"
      color={null}
      inverted={false}
    >
      EAST
    </Segment>
  </GridColumn>
  <GridColumn
    key="SOUTH"
  >
    <Segment
      className="seat-cell"
      color={null}
      inverted={false}
    >
      SOUTH
    </Segment>
  </GridColumn>
</GridRow>
`;

exports[`SeatRow render vulnerability: NS 1`] = `
<GridRow
  className="seat-row"
  column={4}
  key="seat"
>
  <GridColumn
    key="WEST"
  >
    <Segment
      className="seat-cell"
      color={null}
      inverted={false}
    >
      WEST
    </Segment>
  </GridColumn>
  <GridColumn
    key="NORTH"
  >
    <Segment
      className="seat-cell"
      color="red"
      inverted={true}
    >
      NORTH
    </Segment>
  </GridColumn>
  <GridColumn
    key="EAST"
  >
    <Segment
      className="seat-cell"
      color={null}
      inverted={false}
    >
      EAST
    </Segment>
  </GridColumn>
  <GridColumn
    key="SOUTH"
  >
    <Segment
      className="seat-cell"
      color="red"
      inverted={true}
    >
      SOUTH
    </Segment>
  </GridColumn>
</GridRow>
`;
